rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,363 @@
+    default_ttl: Option<Duration>
+}
+```
+
+To enforce that the clock remains monotonically increasing across instantiations of the
+database, we will add the maximum timestamp written in both the manifest and each SST (
+see the change to `SsTableInfo` below). When seeding the database minimum timestamp, SlateDB
+will use the maximum value seen in all persisted WAL SSTs and the latest manifest.
+
+```fbs
+table ManifestV1 {
+    // ...
+    timestamp: long;
+}
+```
+
+`Db#put_put_with_options` can optionally specify a row-level TTL at insertion time
+by leveraging newly added field in `WriteOptions`:
+
+```rust
+pub struct WriteOptions {
+    /// The time-to-live (ttl) for this insertion. If this insert overwrites an existing
+    /// database entry, the TTL for the most recent entry will be canonical.
+    ///
+    /// Default: the TTL configured in DbOptions when opening a SlateDB session
+    pub ttl: Option<Duration>,
+}
+```
+
+For this RFC we will not introduce the ability to disable filtering out TTL-expired records on
+retrieval. The default implementation will read the metadata for every retrieved row and filter out
+records where `ts + tll < now`. We can, as a future optimization, enable lazy deserialization of the
+metadata so that this doesn't need to happen on each row retrieved if the user requests the ability
+to retrieve TTL'd rows that have not yet been compacted. This can be part of the API design for
+compaction filters (see https://github.com/slatedb/slatedb/issues/225)
+
+## Implementation
+
+### Data Format
+
+Currently, SlateDB does not have a mechanism for backwards-compatibility of the SST data format.
+
+We will first introduce the following fields to the flatbuffer definitions:
+
+- `format_version`: field that specifies the codec version that encoded this SST
+- `meta_features`: an enumeration of metadata fields that are available with each encoded row. This
+  makes it possible in the future to save space per row by disabling metadata features, but that is
+  out of scope for this RFC
+- `creation_timestamp`: the time at which this SST file was written. For this RFC this timestamp is
+  used to schedule follow-up compactions (see [periodic compaction](#periodic-compaction)).
+- `min_ts`/`max_ts`:  metadata that indicates the earliest and latest timestamp values in the SST
+
+```fbs
+/// the metadata encoded with each row, note that the ordering of this enum is 
+/// sensitive and must be maintained
+enum MetaFeatures: byte {
+    Timestamp,
+    TimeToLive,
+    RowFlags,
+}
+
+table SsTableInfo {
+    // ...
+    
+    // The current encoding version of the data block
+    format_version: uint;
+    
+    // The metadata features that are encoded with each row. These features will be encoded
+    // in order that they are declared in the MetaFeatures enum
+    meta_features: [MetaFeatures];
+    
+    // The time at which this SST was created
+    creation_timestamp: long;
```

Comment:
Nit: All these timestamps, such as `creation_timestamp`, `min_ts` would be based on the supplied `Clock`. I think making it explicit in some of the comments would be good.